File: server/services/mcp/firewall.js
'use strict';

function decide(toolName, policy = {}) {
  const name = String(toolName || '').trim();
  const failClosed = policy.fail_closed !== false;
  const allowed = new Set(policy.allowed_tools || policy.allowed || []);
  const approval = new Set(policy.approval_required || policy.approve || []);
  const blocked = policy.blocked_patterns || [];

  if (!name) {
    return { decision: 'deny', tool_name: '', reason: 'empty_tool_name' };
  }
  for (const pattern of blocked) {
    try {
      if (new RegExp(pattern).test(name)) {
        return { decision: 'deny', tool_name: name, reason: `blocked_pattern:${pattern}` };
      }
    } catch {
      if (pattern && name.includes(pattern)) {
        return { decision: 'deny', tool_name: name, reason: `blocked_pattern:${pattern}` };
      }
    }
  }
  if (allowed.has(name)) {
    return { decision: 'allow', tool_name: name, reason: 'allowlisted' };
  }
  if (approval.has(name)) {
    return { decision: 'require_approval', tool_name: name, reason: 'approval_required' };
  }
  if (failClosed) {
    return { decision: 'deny', tool_name: name, reason: 'not_allowlisted' };
  }
  return { decision: 'require_approval', tool_name: name, reason: 'fail_open_approval' };
}

module.exports = { decide };

File: python/agentshield/mcp/firewall.py
"""MCP / tool firewall allowlist primitive.

Phase 1 start: deterministic allow / deny / require_approval decisions.
Full IFC and policy DSL land in Phase 2.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Iterable, Mapping


class FirewallDecision(str, Enum):
    ALLOW = "allow"
    DENY = "deny"
    REQUIRE_APPROVAL = "require_approval"


@dataclass(frozen=True)
class FirewallResult:
    decision: FirewallDecision
    tool_name: str
    reason: str

    def to_dict(self) -> dict[str, str]:
        return {
            "decision": self.decision.value,
            "tool_name": self.tool_name,
            "reason": self.reason,
        }


class ToolAllowlist:
    """Least-privilege tool gate.

    - `allowed`: tools that may run without extra approval
    - `approval_required`: tools that need human-in-the-loop
    - everything else is denied when `fail_closed=True` (production default)
    """

    def __init__(
        self,
        allowed: Iterable[str] | None = None,
        approval_required: Iterable[str] | None = None,
        *,
        fail_closed: bool = True,
    ) -> None:
        self.allowed = {n.strip() for n in (allowed or []) if n and n.strip()}
        self.approval_required = {
            n.strip() for n in (approval_required or []) if n and n.strip()
        }
        self.fail_closed = fail_closed

    def decide(self, tool_name: str) -> FirewallResult:
        name = (tool_name or "").strip()
        if not name:
            return FirewallResult(
                FirewallDecision.DENY, tool_name="", reason="empty_tool_name"
            )
        if name in self.allowed:
            return FirewallResult(FirewallDecision.ALLOW, name, "allowlisted")
        if name in self.approval_required:
            return FirewallResult(
                FirewallDecision.REQUIRE_APPROVAL, name, "approval_required"
            )
        if self.fail_closed:
            return FirewallResult(FirewallDecision.DENY, name, "not_allowlisted")
        return FirewallResult(
            FirewallDecision.REQUIRE_APPROVAL, name, "fail_open_approval"
        )

    def check_call(self, tool_call: Mapping[str, object]) -> FirewallResult:
        name = str(tool_call.get("name") or tool_call.get("tool") or "")
        return self.decide(name)
